onnxruntime/core/session/provider_registration.cc

```diff
@@ -74,7 +75,15 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsAppendExecutionProvider,
 #endif
   } else if (strcmp(provider_name, "XNNPACK") == 0) {
 #if defined(USE_XNNPACK)
+    // If Xnnpack is enabled and user don't give a specific value, we need to set it same as cpu threadpool size
+    if (provider_options.find("thread_num") == provider_options.end()) {
+      provider_options["thread_num"] = std::to_string(options->value.intra_op_param.thread_pool_size);
+    }
     options->provider_factories.push_back(XnnpackProviderFactoryCreator::Create(provider_options));
+    // XNNPACK owns its own threadpool, so we have to disable ort-threadpool's spinning feature
+    // otherwise it will cause contention.
+    status = onnxruntime::ToOrtStatus(options->value.config_options.AddConfigEntry(
+        kOrtSessionOptionsConfigAllowIntraOpSpinning, "0"));
```

Comment:
Then want if user needs ort threadpool and its spinning for heavy ops on cpu EP?

Reply:
I believe this is for mobile scenarios where spinning eats up all the battery.

Reply:
Agree with pranavsharma, we should either implement all heavy ops or disable XNNPACK EP. Ort threadpool will also eats up half of CPUs even it's in IDLE with spinning. 
A workaround is to insert a spinning switch operation when switch from an EP to another EP, the operation should turn off the former's threadpool spinning and turn on the current threadpool spinning.

Reply:
Let's assume for XNNPACK to be a viable option it has kernels for all heavy ops. User can enable/disable based on perf testing.

I not sure we want to set any defaults for now as I don't think we know enough to know what is best in general, as that might be heavily dependent on the model and device. 

We should clearly document that an XNNPACK EP user should try two things. 
1) Set SessionOptions intra_op_num_threads to 1 and the xnnpack threadpool size to the number of cores (or whatever value they would have used for intra_op_num_threads if xnnpack wasn't enabled)
2) Set xnnpack threadpool size to the same size as intra_op_num_threads and disable spinning.

If we want to be really pro-active we could check if they did either of these things, and output a warning telling them to do so (message should point to doco so they understand the pros/cons of the two choice). 

The check might be something like there's no provider option for xnnpack EP setting number of threads, or ORT intra_op_num_threads > 1 and the disable spinning config option isn't set, output warning message.

Reply:
Either way, spinning should be disabled for better performance and lower power consumption.
We could warn something where config is not correct or not appropriate.
But we could do something like "always disable spinning", is that right?